Generating absolute prestate and preimage files for permissionless proofs (#1610)
ethereum-optimism/docs: pages/operators/chain-operators/tutorials/absolute-prestate.mdx

*   Support for the latest network upgrades

<Callout type="info">
  Beginning with [Upgrade 14](/notices/upgrade-14), all chains should use the `64-bit` multi-threaded version of Cannon. 
  The absolute prestate files for this version typically have the format `prestate-mt64.bin.gz`.
</Callout>

## Generating the absolute prestate

<Steps>
  ### Clone and checkout the tagged version

  First, clone the Optimism monorepo and check out the appropriate [release tag](https://github.com/ethereum-optimism/optimism/tags) for op-program:

  ```bash
  git clone https://github.com/ethereum-optimism/optimism.git
  cd optimism
  git checkout op-program/v1.6.1-rc.1  # Use the latest tagged version
  git submodule update --init --recursive

```

<Callout type="important">
    Always use a specific tagged version of op-program as specified in the Superchain registry or official release notes. Current releases use versions like `op-program/v1.6.0-rc.2` for Isthmus hardfork support.
  </Callout>

### Build the op-program

Build the op-program binary:

```bash
make op-program
```

### Generate the absolute prestate

Generate the prestate by running the op-program binary directly:

```bash

> bash

Reviewer: remove bash

./op-program/bin/op-program prestate

> 

Reviewer: I'm not sure this is the right command. Seems to me this takes l1.head and l2.claim as inputs and checks if the L2 state matches what is expected according to the L1.

```

> For older versions (prior to Upgrade 14), you might use:

Reviewer: Add a note to make sure Docker is running at this point.

Author: Added this.

```bash
cd op-program

> make reproducible-prestate

Reviewer: This seems to work fine also with the version checked out at the beginning of the doc. Might still be relevant and the correct command also after U14, I suggest to double check this.

## Prerequisites

Before starting, ensure you have:

*   [Docker](https://docs.docker.com/engine/install/) running
*   Have sufficient disk space and memory

## Using the latest absolute prestate and multi-threaded Cannon

As of the latest releases, OP Stack chains should utilize the `64-bit` MIPS multi-threaded version of Cannon. 
This upgrade offers several advantages:

*   Improved performance through parallelized execution
*   Enhanced fault proof VM capabilities
*   Support for the latest network upgrades

<Callout type="info">
  Beginning with [Upgrade 14](/notices/upgrade-14), all chains should use the `64-bit` multi-threaded version of Cannon. 
  The absolute prestate files for this version typically have the format `prestate-mt64.bin.gz`.
</Callout>

## Generating the absolute prestate

<Steps>
  ### Clone and checkout the tagged version

  First, clone the Optimism monorepo and check out the appropriate [release tag](https://github.com/ethereum-optimism/optimism/tags) for op-program:

  ```bash
  git clone https://github.com/ethereum-optimism/optimism.git
  cd optimism

>   git checkout op-program/v1.6.1-rc.1  # Use the latest tagged version

Reviewer 2: For production chains you would normally want the latest finalized release (e.g. `v1.6.1` not an rc).  The latest finalized release is typically what is running on op-mainnet - testnets and devnets will use rc's.

  git submodule update --init --recursive # Initialize submodules

```

### Build the op-program

Build the op-program binary:

```bash
make op-program

> ```

Reviewer 2: This shouldn't be required. `make reproducible-prestate` will build everything necessary.

### Generate the absolute prestate

Generate the prestate by running the op-program binary directly:

```bash

> cd op-program
> make reproducible-prestate

Reviewer 2: No need to cd op-program here. `make reproducible-prostate` is a target in the top level makefile.
```suggestion
make reproducible-prestate
```

cd ..
```

The output should look like this:

```bash

-------------------- Production Prestates --------------------

Cannon64 Absolute prestate hash: 
0x03eb07101fbdeaf3f04d9fb76526362c1eea2824e4c6e970bdb19675b72e4fc8

-------------------- Experimental Prestates --------------------

CannonInterop Absolute prestate hash: 
0x03fc3b4d091527d53f1ff369ea8ed65e5e17cc7fc98ebf75380238151cdc949c

Cannon64Next Absolute prestate hash: 
0x03eb07101fbdeaf3f04d9fb76526362c1eea2824e4c6e970bdb19675b72e4fc8

```

  The output should display three prestate hashes:

  *   `Cannon64`: The current production absolute prestate hash for the 64-bit version of Cannon
  *   `CannonInterop`: The absolute prestate hash used for interoperability
  *   `Cannon64Next`: The next state version for Cannon with in-progress features (like Go 1.24 support)

>   For permissionless fault proofs, you'll primarily use the Cannon64 absolute prestate hash.

Reviewer 2: The actual output will vary as we change what versions are available. I would probably just explain the Cannon64 prestate which appears in production prestates and note that other prestates will be listed in the experimental section but they aren't ready for use yet.

  ### Prepare the preimage file

>   After generating the prestate, find the preimage file.
>   You should see a file named `prestate-mt64.bin.gz` or similar (the exact name might vary based on the version).
>   Rename this file to include the prestate hash:
>   The preimage file is located at `op-program/bin/prestate-mt64.bin.gz`. Rename it to match the prestate hash:

Reviewer 2: ```suggestion
  After generating the prestate, the preimage file will be located in `op-program/bin/prestate-mt64.bin.gz`. The exact name might vary based on the version.
  Rename this file to include the prestate hash:
```

  ```bash
  cd op-program/bin
  mv prestate-mt64.bin.gz 0x[CANNON64_PRESTATE_HASH].bin.gz
  ```

  Replace `[CANNON64_PRESTATE_HASH]` with the actual `Cannon64` absolute prestate hash value from the output.
  This file needs to be uploaded to a location that's accessible by your op-challenger instances.
</Steps>

## Official prestate hashes for Superchain registry chains

The Superchain registry maintains official absolute prestate hashes for chains that are part of the registry. 

> These prestates include the configurations of all chains in the Superchain registry.

Reviewer 2: This isn't necessarily true. The prestates contain the config for chains that were in the version of the registry imported to the monorepo when the prestate was tagged. Just because a prestate is listed in the superchain-registry doesn't mean it is safe to use for a chain in the registry (it may have been created before the chain was added or it may not contain the most up to date config for the chain).

Reviewer 2: We have a very hacky little tool for checking if a prestate listed in the superchain registry includes the latest config for a chain:
https://github.com/ethereum-optimism/optimism/blob/273fd6abd4a92824e05f3daa9322b0c8add9e843/op-chain-ops/cmd/check-prestate/README.md#L1

It's just a quick tool for us and not very user friendly.

For chains listed in the Superchain registry, you should use the official prestate hashes rather than generating your own.

You can find the latest prestate tags in the [Superchain registry](https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/standard/standard-prestates.toml).

## Handling unannounced chain configurations

If your chain is not included in the Superchain registry, you'll need to generate a custom prestate with your specific chain configuration. This applies to:

*   New chains
*   Chains with custom configurations

For these cases, follow these additional steps:

<Steps>

### Create a directory for your custom chain configuration
```bash

> mkdir -p op-program/chainconfig/configs

Reviewer 2: This directory will already exist in a valid checkout of the monorepo (it has a `placeholder.json` to ensure it always exists). If it doesn't already exist you're probably on a version that doesn't support custom configs.

```

### Copy your chain configuration files to this directory

First, you need to obtain your chain's configuration files. These are typically generated when you [deploy your L2 chain](/operators/chain-operators/deploy/smart-contracts) with op-deployer:

*   ***rollup.json**: The rollup configuration file
*   **genesis.json**: The L2 genesis file 

Name them according to the required format:

```bash
# Replace 67865 with your actual chain ID
cp /path/to/rollup.json op-program/chainconfig/configs/67865-rollup.json
cp /path/to/genesis.json op-program/chainconfig/configs/67865-genesis-l2.json
```
Note: The naming format is critical - the files must be named as:

*   `<L2-chain-id>-rollup.json`
*   `<L2-chain-id>-genesis-l2.json`

> These files are outputs from your L2 chain deployment process using op-deployer.

Reviewer 2: It is important that the prestate remains available so it is recommend to commit these config files to a git branch that you keep in your own fork of the monorepo.  Then you can rebuild the prestate if needed by checking out that fork and running `make reproducible-prestate`.

### Generate the prestate

From the root of the monorepo, run:

```bash
make reproducible-prestate
```
Alternatively, you can also specify the paths directly:

```bash

> make reproducible-prestate ROLLUP_CONFIG=/path/to/67865-rollup.json L2_GENESIS=/path/to/67865-genesis-l2.json

Reviewer 2: This doesn't work.  You _have_ to put the files in the config directory and then just run `make reproducible-prestate`.

0x03eb07101fbdeaf3f04d9fb76526362c1eea2824e4c6e970bdb19675b72e4fc8

```

For current production use, you should use the `Cannon64` Absolute prestate hash.

### Locate and prepare the preimage File
Follow this [step](/operators/chain-operators/tutorials/absolute-prestate#prepare-the-preimage-file)

</Steps>

## Deploying and configuring with the absolute prestate

After generating the absolute prestate and preimage files, you'll need to:

<Steps>
  ### Upload preimage file

  Upload the preimage file to a location accessible by your op-challenger instances

  ### Configure op-challenger

  Configure the op-challenger to use this prestate:

  ```bash
  docker run -d --name op-challenger \
    -e OP_CHALLENGER_TRACE_TYPE=permissioned,cannon \
    -e OP_CHALLENGER_PRESTATES_URL=<YOUR_PRESTATES_URL> \
    -e OP_CHALLENGER_L1_ETH_RPC=<YOUR_L1_RPC_URL> \
    -e OP_CHALLENGER_GAME_FACTORY_ADDRESS=<YOUR_DISPUTE_GAME_FACTORY> \
    -e OP_CHALLENGER_PRIVATE_KEY=<YOUR_PRIVATE_KEY_OR_USE_WALLET_SIGNER> \
    -e OP_CHALLENGER_NETWORK=<YOUR_NETWORK> \
    -e OP_CHALLENGER_CANNON_ROLLUP_CONFIG=<PATH_TO_ROLLUP_CONFIG> \
    -e OP_CHALLENGER_CANNON_L2_GENESIS=<PATH_TO_GENESIS_CONFIG> \

>     -v /path/to/local/prestates:/prestates \

Reviewer 2: Why is this mount required? If `YOUR_PRESTATES_URL` is a http URL you don't need to mount the prestates as a version.  Otherwise if you do mount them you could use a file URL and it would just be `-e OP_CHALLENGER_PRESTATES_URL=file:///prestates` with this mount definition.

    us-docker.pkg.dev/oplabs-tools-artifacts/images/op-challenger:latest #use the latest version
  ```

  Replace `<URL_TO_PRESTATES_DIRECTORY>` with the URL where you've stored your prestate files.

>   The `--prestates-url` should point to the directory where you've uploaded the renamed prestate file from step 3.

Reviewer 2: There is no `<URL_TO_PRESTATES_DIRECTORY>` in the above command. I think you mean `<YOUR_PRESTATES_URL>`.  `--prestates-url` must be a URL and since it's running in docker it would have to point to the mount point, not the directory on the host where the prestate was built.
